### PR TITLE
Updated code to correctly minify the library with JSCompiler.

### PR DESCRIPTION
**Pull request by TMSCH**
This required adding externs that include type definitions for interfaces that are exchanged between the client and the provider. 
Updated the travis script to make sure it compiles correctly.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -33,8 +33,12 @@
     "compile": "yarn run compile-dev && yarn run compile-demos",
     "prerollup-api": "yarn run compile-es6",
     "rollup-api": "./node_modules/.bin/rollup -c rollup.api.config.js",
+    "postrollup-api": "java -jar node_modules/google-closure-compiler/compiler.jar --compilation_level ADVANCED_OPTIMIZATIONS --js es5/openyolo-api.iife.js --js_output_file es5/openyolo-api.iife.min.js --externs externs/openyolo.js",
     "prerollup-spi": "yarn run compile-es6",
     "rollup-spi": "./node_modules/.bin/rollup -c rollup.spi.config.js",
+    "postrollup-spi": "java -jar node_modules/google-closure-compiler/compiler.jar --compilation_level ADVANCED_OPTIMIZATIONS --js es5/openyolo-spi.iife.js --js_output_file es5/openyolo-spi.iife.min.js --externs externs/openyolo.js",
+    "gzip-size-api": "echo 'API:' && gzip -c es5/openyolo-api.iife.min.js | wc -c",
+    "gzip-size-spi": "echo 'SPI:' && gzip -c es5/openyolo-spi.iife.min.js | wc -c",
```
Comment (tikurahul) [inline package.json]: Awesome :+1:

File: package.json [hunk 2]
```diff
@@ -44,7 +48,7 @@
     "lint-core-fix": "./node_modules/.bin/tslint --project tsconfig.json --fix",
     "lint-demo": "./node_modules/.bin/tslint --project demo/tsconfig.json",
     "lint-demo-fix": "./node_modules/.bin/tslint --project demo/tsconfig.json --fix",
-    "prepublish": "yarn run compile-es5 && yarn run compile-es6 && yarn run rollup-api && yarn run rollup-spi",
+    "prepublish": "yarn run compile-es5 && yarn run compile-es6 && yarn run rollup-api && yarn run rollup-spi && npm run gzip-size-api && npm run gzip-size-spi",
```
Comment (tikurahul) [inline package.json]: :1st_place_medal:

File: ts/protocol/enums.ts [hunk 3]
```diff
@@ -32,6 +32,11 @@ export function strEnum<T extends string>(...enumValues: T[]):
   return Object.freeze(result);
 }
 
+export function map2Enum<T extends string>(mapValues: {[K in T]: K}):
+    Readonly<{[K in T]: K}> {
```
Comment (tikurahul) [inline ts/protocol/enums.ts]: You might be able to use a `ReadOnlyMap`.

File: ts/api/api.ts [hunk 4]
```diff
@@ -362,3 +362,21 @@ class InitializeOnDemandApi implements OnDemandOpenYoloApi {
 }
 
 export const openyolo: OnDemandOpenYoloApi = new InitializeOnDemandApi();
+
+// Export the public methods.
+const windowAsAny = window as any;
+windowAsAny['openyolo'] = openyolo;
+InitializeOnDemandApi.prototype['setProviderUrlBase'] =
```
Comment (iainmcgin) [inline ts/api/api.ts]: For these, we found that as tsc preserves comments you could use a JSDoc comment with @export on each method that needs to be preserved. That would at least seem cleaner than these no-op redeclarations.

File: ts/protocol/comms.ts [hunk 5]
```diff
@@ -71,14 +71,14 @@ export function createMessageListener<T extends MessageType>(
       console.debug('non openyolo message received');
       return false;
     }
-    if (ev.data.type !== type) return false;
+    if (ev.data['type'] !== type) return false;
```
Comment (iainmcgin) [inline ts/protocol/comms.ts]: it's annoying that closure is forcing you to do this; does it trip up lint?

File: ts/api/api.ts (hunk 4 above)
Comment (TMSCH) [inline ts/api/api.ts]: tsc does preserve comments, but the rollup phase deletes them... So it seems to be the only solution for now. And it's actually the method recommended by the Closure doc: https://developers.google.com/closure/compiler/docs/api-tutorial3#export.

File: ts/protocol/comms.ts (hunk 5 above)
Comment (TMSCH) [inline ts/protocol/comms.ts]: For these parts where messaging between the two libraries is involves, I'm using this bracket notations to make sure it's not renamed in the minification phase. I will revisit that once we have E2E tests that ensure that the two minified libraries communicate properly.

**Comment by TMSCH** [comment]
Thanks Rahul and Iain!

**Pull request merged**